### PR TITLE
Basic caching with time-based expiration

### LINKED ISSUE #15: Persist/Cache Stats downloaded data
Original issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/1555

If Simperium will eventually back Stats data, we need to persist the data in Core Data rather than being in memory only.  This will also go to support viewing the screens in an offline / transient state instead of showing errors or a painted screen over saying nothing here.

Optionally, provide a caching mechanism that can be ported into the future to persist into Core Data.

### PR DESCRIPTION
**Pull request by astralbodies**
Closes #15 

Adds a basic caching mechanism to the WPStatsService. I chose to include the cache `StatsEphemory` at the service layer because it's a form of a persistence mechanism. The implementation detail should not be part of the view controller. I feel like there is room for improvement over the way I've implemented it but it's a simple solution to start with.

At some point having a layer in front of the service makes sense because it can cache stats for multiple sites and persist beyond the somewhat disposable instance of `WPStatsService`.

Pull to refresh forces the cache to evict all items as it's a user-requested force refresh. Tapping between periods and dates will not force a refresh.

cc: @jleandroperez

### CODE REVIEW

File: WordPressCom-Stats-iOS/StatsEphemory.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#import <Foundation/Foundation.h>
+
+@interface StatsEphemory : NSObject
```
Comment (SergioEstevao) [inline WordPressCom-Stats-iOS/StatsEphemory.h]: I like the name but for for future developers that need to maintain the code in the future, StatsCache will be more obvious no?

File: WordPressCom-Stats-iOS/WPStatsService.m [hunk 2]
```diff
@@ -56,26 +85,87 @@ - (void)retrieveAllStatsForDates:(NSArray *)dates
...
+    } else {
+        cacheDictionary = [NSMutableDictionary new];
+        [self.ephemory setObject:cacheDictionary forKey:@[@(unit), endDate]];
     }
+
+    [self.remote batchFetchStatsForDate:endDate
+                                andUnit:unit
+            withVisitsCompletionHandler:^(StatsVisits *visits, NSError *error)
+     {
+         cacheDictionary[@(StatsCacheVisits)] = visits;
```
Comment (SergioEstevao) [inline WordPressCom-Stats-iOS/WPStatsService.m]: if you get an error on the remote call, will the visits be nil and then the cache insertion fail?
You may need to check for error first before doing this.

This happens for all completion handlers but I will mention only this time.

File: WordPressCom-Stats-iOS/WPStatsService.m [hunk 3]
```diff
@@ -227,8 +339,29 @@ - (void)retrieveTodayStatsWithCompletionHandler:(void (^)(StatsSummaryCompletion
         }
     };
     
+    if (!completion) {
+        return;
+    }
+    
+    StatsSummary *summary = [self.ephemory objectForKey:@"TodayStats"];
```
Comment (jleandroperez) [inline WordPressCom-Stats-iOS/WPStatsService.m]: This is me being annoying: Maybe a const for @"TodayStats"?

**Comment by jleandroperez** [comment]
@astralbodies Code wise it looks very very nice!.

Once the comments are addressed, i'd say :shipit: !

File: WordPressCom-Stats-iOS/StatsEphemory.h (hunk 1 above)
Comment (astralbodies) [inline WordPressCom-Stats-iOS/StatsEphemory.h]: Added some header documentation to accommodate the obscurity of the name. :smile:

File: WordPressCom-Stats-iOS/WPStatsService.m (hunk 2 above)
Comment (astralbodies) [inline WordPressCom-Stats-iOS/WPStatsService.m]: The remote layer always returns an object. I can change this in the future. I did fix a missing BOOL flip if there was an error.

File: WordPressCom-Stats-iOS/WPStatsService.m (hunk 3 above)
Comment (astralbodies) [inline WordPressCom-Stats-iOS/WPStatsService.m]: This method isn't done yet, more of a stub. I'll definitely make it a constant in that PR :smile: (I was just tired of the warning of the unused failure handler :smiley_cat:)

File: WordPressCom-Stats-iOS/StatsEphemory.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#import <Foundation/Foundation.h>
+
+/*!
+    @class      StatsEphemory
+    @brief      A lightweight container for NSCache.
+    @discussion Encapsulates an NSCache instance with an additional parameter for expiration time interval.
+ 
+                This class' name is a portmanteua of Ephemeral and Memory.
```
Comment (sendhil) [inline WordPressCom-Stats-iOS/StatsEphemory.h]: SAT words!

**Pull request merged**